mds: scrub abort/pause/resume control operations

vshankar opened the pull request:
This PR is a part of the much wider "show scrub info in ceph -s" tracker (http://tracker.ceph.com/issues/36370) and is WIP at the moment as some other operations (such as scrub progress) is in midst of being implemented w/ as part of the above tracker.

vshankar commented:
@batrick : w.r.t. http://tracker.ceph.com/issues/36370, would it be better to have a separate mds asok command to display scrub information (rather than it being a part ceph cluster status output). With this PR, there is already a `scrub status` asok command that displays current scrub status (active, paused -- w/ number of inodes in the stack waiting to be scrubbed). We could probably extend that command to show ongoing scrub information (i.e., arguments to `scrub_path` command such as subpath and other control flags).

I think having a separate cli for this functionality would be better in many ways. Wanted to know your opinion on this.

batrick commented:
@ukernel please review

ukernel commented on src/mds/ScrubStack.cc:
```diff
@@ -436,3 +498,140 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+  if (state == STATE_NOT_RUNNING) {
+    *ss << "no active scrubs running";
+  } else if (state == STATE_RUNNING) {
+      *ss << "scrub active (" << stack_size << " inodes in the stack)";
```
I think clear_inode_stack can be true in running state.

vshankar replied:
you are correct -- I'll rearrange the status part.

regarding having clear_inode_stack as a bit in a state (STATE_RUNNING), I think, if there are other possible operations that have to be performed in STATE_RUNNING, keeping a bitmap of operations would be useful. Moreover, it would be more useful if we can keep a `state -> operation bitmap` mapping to make it more generic. Right now, we just have abort operation, so using a bool does the job.

vshankar commented:
Keeping this WIP since I haven't included the `scrub progress` part with this update.

* scrub control commands use the `tell` interface. This also tempted me to move the existing `scrub_path` and `tag path` to use `tell` interface (for uniformity).
* `scrub_path` now generates a random tag (uuid). `scrub status` displays running scrubs with their respective tags.

vshankar commented:
Removed WIP from the description -- this is pretty much reviewable.

Also, I realized that I had added progress (n dentries out of m scrubbed) as part of `scrub status` output, but didn't push it as an update. I think this needs a bit more thought for such information to be useful since files/directories can be created when the scrub is ongoing.

Can the progress bits be targetted as a followup PR?

ukernel commented on src/mds/MDCache.cc:
```diff
@@ -12217,6 +12217,15 @@ void MDCache::enqueue_scrub(
+    uuid_gen.print(uuid_str);
+    tag = uuid_str;
+  }
+
```
non-empty scrub tag will get stored in inodes' xattr.  This change will introduce unnecessary writes

vshankar replied:
the tag will be persisted for every inode in the subtree being scrubbed, right? we could skip this IO operation if the tag is internal and not provided via CLI.

ukernel replied:
yes, current code store tag for every inodes.  yes, we could modify the code to not store internal tag

batrick commented on src/mds/MDSDaemon.cc:
```diff
@@ -664,6 +664,13 @@ const std::vector<MDSDaemon::MDSCommand>& MDSDaemon::get_commands()
     MDSCommand("cache drop name=timeout,type=CephInt,range=0,req=false", "trim cache and optionally request client to release all caps and flush the journal"),
+    MDSCommand("scrub_path name=path,type=CephString name=scrubops,type=CephChoices,strings=force|recursive|repair,n=N,req=false",
+               "scrub an inode and output results"),
+    MDSCommand("tag path name=path,type=CephString name=tag,type=CephString", "Apply scrub tag recursively"),
```
`tag path` appears to be an odd historical artifact. I don't think we should add this command since it's just `scrub_path` with hard-coded options.

batrick commented on src/mds/MDSDaemon.cc:
```diff
@@ -664,6 +664,13 @@ const std::vector<MDSDaemon::MDSCommand>& MDSDaemon::get_commands()
         "name=heapcmd,type=CephChoices,strings=dump|start_profiler|stop_profiler|release|stats",
         "show heap usage info (available only if compiled with tcmalloc)"),
     MDSCommand("cache drop name=timeout,type=CephInt,range=0,req=false", "trim cache and optionally request client to release all caps and flush the journal"),
+    MDSCommand("scrub_path name=path,type=CephString name=scrubops,type=CephChoices,strings=force|recursive|repair,n=N,req=false",
```
Let's be consistent by using spaces, so `scrub path`. Or, even better, we could adopt `scrub start` so it's harmonious with the new scrub commands you're adding below.

vshankar replied:
I kept the command as-it-is -- I'm ok with `scrub start` since its much clear.

My only worry is that the asok interface is `scrub_path` and tell interface would be `scrub start` -- shouldn't the commands be consistent? (the asok command can be flipped to `scrub start` too)...

batrick replied:
I think the asok commands should just become deprecated.

batrick commented on src/mds/ScrubStack.h:
```diff
@@ -76,16 +80,69 @@ class ScrubStack {
+
 private:
+  // ABORTING is _not_ a state, rather it's an operation that's
+  // performed after in-progress scrubs are finished.
```
I find this comment confusing. It sounds like it's trying to explain a `State::STATE_ABORTING` but no state exists.

vshankar replied:
so, looks like `ABORTING` vs `aborting` caused the confusion :)

I'll clear that out...

batrick commented on src/mds/ScrubStack.h:
```diff
@@ -76,16 +80,69 @@ class ScrubStack {
+  bool clear_inode_stack = false;
+  std::list<Context *> control_ctxs;
+
+  std::set<CInode *> scrub_origins;
```
Please explain what this is for.

batrick commented on src/mds/ScrubStack.h:
```diff
@@ -76,16 +80,69 @@ class ScrubStack {
+
+  State state = STATE_NOT_RUNNING;
+  bool clear_inode_stack = false;
+  std::list<Context *> control_ctxs;
```
Please add a comment explaining what this is for.

batrick commented on src/mds/ScrubStack.cc:
```diff
@@ -436,3 +501,191 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+void ScrubStack::complete_control_contexts(int r) {
+  ceph_assert(mdcache->mds->mds_lock.is_locked_by_me());
+
+  for (auto ctx : control_ctxs) {
```
`auto& ctx`. (Pointer copies are cheap but this makes the reader wonder.)

batrick commented on src/mds/ScrubStack.cc:
```diff
@@ -436,3 +501,191 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+  f->dump_string("status", ss.str());
+
+  f->open_object_section("scrubs");
+  for (auto inode : scrub_origins) {
```
`auto& inode`

batrick commented on src/mds/ScrubStack.cc:
```diff
@@ -436,3 +501,191 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
+    r = -EINVAL;
+  } else if (state == STATE_PAUSING) {
+    set_state(STATE_RUNNING);
+    complete_control_contexts(-ECANCELED);
```
Can you explain this error code? Is this trying to indicate failure to a running `scrub pause`?

vshankar replied:
This is explained in the comment here: https://github.com/ceph/ceph/pull/24635/commits/b720fdadc7237677a21cf06313687c1e4ec64b60#diff-1692e6ddf1d8f2b221b91048bbacf70bR116

In short, resuming a pending pause scrub operation will indicate the caller (via context finisher) that the pause operation has been canceled.

batrick replied:
ACK

batrick commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -11,6 +11,108 @@
+
+    def _abort_scrub(self, expected):
+        mds_id = self.fs.get_lone_mds_id()
+        res = json.loads(self.fs.mon_manager.raw_cluster_cmd("tell", "mds.{0}".format(mds_id), "scrub", "abort"))
```
Let's define a Filesystem method that gets rank 0's id and executes ceph tell with the given command.

vshankar commented on src/mds/MDSRank.cc:
```diff
@@ -2443,13 +2443,23 @@ bool MDSRankDispatcher::handle_asok_command(std::string_view command,
+    C_SaferCond cond;
+    {
+      command_scrub_path(f, path, scrubop_vec, &cond);
+    }
```
will remove -- used to have some local var. thanks for catching.

vshankar commented:
> @vshankar looks like we're still getting the async output even when running the sync version?

will take a look...

vshankar commented:
> @vshankar looks like we're still getting the async output even when running the sync version?

nope -- recursive scrub is triggered for the given failed test. The reason why the test failure is due to the reason of dumping additional data when initiating scrub (`tag` and `mode`) w/ the test validating an empty json or a valid json having `return_code` as a member.

vshankar commented:
@batrick fixed and updated -- included `return_code` in json output for recursive scrub and switched cephfs upgrade snap test to use `scurb start` tell interface (plus a minor doc change to use `scrub start`).

reran the tests that failed here -- http://pulpito.ceph.com/vshankar-2019-01-02_12:17:51-fs-wip-scrub-control-distro-basic-mira/
